jib-core/src/main/java/com/google/cloud/tools/jib/configuration/ImageConfiguration.java

```diff
@@ -30,6 +34,8 @@
 
     private ImageReference imageReference;
     private ImmutableList<CredentialRetriever> credentialRetrievers = ImmutableList.of();
+    @Nullable private DockerClient dockerClient;
+    @Nullable private Path tarPath;
```

Comment:
I have had some reservation, as this feels a bit like a quick ad-hoc fix and doesn't feel like OO-style. `dockerClient` and `tarPath` are incompatible in that they shouldn't be both set, and as such, we need some code to guard it. And relying on on the assumption that a non-null `dockerClient` implies the semantic that this `ImageConfiguation` represents a local Docker image (and likewise that a non-null `tarPath` implies a tarball image) whereas a null value implies a remote registry image seems requiring such external domain knowledge in understanding the code. I wish given an image configuration it clearly presented what the image is. Though I can't think of a good and concrete suggestion right now. @GoogleContainerTools/java-tools-build thoughts?

Reply:
I guess we could make `RegistryImage`, `DockerDaemonImage`, and `TarImage` implement a common `BaseImage` interface, store a `BaseImage` in `ImageConfiguration`, and check the type/cast to get the required fields in `StepsRunner`?

Reply:
I'm not sure how this affects the jib-core public API though.

Reply:
In fact if we do this, I think we can remove `ImageConfiguration` completely. In which case I think I'd rather refactor this in a follow-up.

Reply:
Does setting `tarPath` mean that `imageReference` can be optional too?

Reply:
Yes for base images, no for target images. It's handled in a somewhat hacky way [here](https://github.com/GoogleContainerTools/jib/pull/1919/files#diff-d390f64e06290a875d4e6a149d0f18ccR98) that I'd like to improve once this part is figured out.

Reply:
> In fact if we do this, I think we can remove ImageConfiguration completely, but that can be handled in a follow-up.

I think I would prefer we designed removing `ImageConfiguration` first and then submitted this PR.

Reply:
What I had in mind for removing `ImageConfiguration` follows my common interface idea; if each image type implemented a common `ContainerImage` interface, then `ImageConfiguration` could just contain that instead of the 4 fields we have now. And at that point, it's unnecessary, so we could remove it and just stick the `ContainerImage` straight into `BuildConfiguration`, since those will always have the fields we need (after casting to the correct type first, at least).

I didn't want to remove it in this PR because it's used in a lot of places and would probably make this PR a lot bigger, and also I'm not sure if the check/cast way of going about this is better than what we have now.

Reply:
I'm fine with taking care of this after merging this. And I still have no good suggestion as how exactly.